### PR TITLE
Implement ==/hashCode for ViewConfiguration, avoid unnecessary layer creation/replacement

### LINKED ISSUE #81514: Establish a lifecycle for `Picture` objects.
`Picture` objects retain anything drawn into them. This can be troubling when they retain `Image` or other large chunks of memory.

Unfortunately, today, `Picture` objects have no lifecycle. They have a `dispose` method that the framework never calls. They are mainly held by `Layer` objects in the framework, which also have no clearly defined lifecycle. They are passed back into the engine side and passed around as shared pointers.

Generally speaking, it requires a garbage collection to figure out when a picture object can be reclaimed. However, this can be expensive and may not happen at an ideal time. Even scheduling collections more aggressively when we believe we're done with an image is imperfect, since it can schedule expensive collections too often.

It would be better if we disposed picture objects at some point where we could be convinced we are done with them, which would allow us to release the native pointers into Skia even if a GC has not occurred. 

The main challenge with this is around `Layer` objects. Layers have no clear owner and get passed around. Additionally, we have three parallel layer trees:

- `flutter::Layer` objects that are meant to be used on the raster task runner by the rasterizer.
- `flutter::EngineLayer` objects which are native peers to Dart `EngineLayer` objects and reference `flutter::Layer` objects.
- `Layer` objects in the Flutter framework, some of which hold `EngineLayer` objects.

~~On the engine side, we should make sure that any Dart UI C++ backing objects drop native references at appropriate times. The only obvious miss right now is `SceneBuilder::build`, which does not let go of the `layer_stack_` ivar, and `Scene::dispose` (which is less serious because typically the scene gets its layer pointer stolen away when it is rendered). ~~

The real issue here is to make sure that the framework does not hold references to `flutter::Layer` objects any longer than necessary, to avoid cleaning up the layer tree via GC.

/cc @yjbanov because we've also talked about how this could help on Safari.

### PR DESCRIPTION
**Pull request by dnfield**
Part of a quest to make layer usage more sensible, to resolve https://github.com/flutter/flutter/issues/81514

Without this change, we unnecessarily replace/recreate the root layer every time we get an update about view configuration, even if the configuration has not actually changed.

I promise I read and fulfilled all the boxes, I'm just lazy about adding x's all over

### CODE REVIEW

File: packages/flutter/lib/src/rendering/view.dart [hunk 1]
```diff
@@ -38,6 +38,18 @@ class ViewConfiguration {
     return Matrix4.diagonal3Values(devicePixelRatio, devicePixelRatio, 1.0);
   }
 
+  @override
+  bool operator==(Object other) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/view.dart:42]: nit: space between `operator` and `==`?

File: packages/flutter/test/rendering/view_test.dart [hunk 2]
```diff
@@ -2,12 +2,23 @@
...
 
+import 'dart:ui' as ui;
+
 import 'package:flutter/rendering.dart';
 import 'package:flutter_test/flutter_test.dart';
 
 import 'rendering_tester.dart';
 
 void main() {
+  // Create non-const instances, otherwise tests pass even if the
```
Comment (goderbauer) [inline packages/flutter/test/rendering/view_test.dart:13]: add a test case for this to encode that two ViewConfiguration returned by this are not `identical`? (could be part of the `ViewConfiguration == and hashCode` test)

File: packages/flutter/test/rendering/view_test.dart [hunk 3]
```diff
@@ -17,5 +28,30 @@ void main() {
...
+      view.configuration = createViewConfiguration(devicePixelRatio: 5.0);
+      expect(identical(view.debugLayer, firstLayer), false);
+    });
+  });
+
+  test('ViewConfiguration == and hashCode', () {
+    final ViewConfiguration viewConfigurationA = createViewConfiguration();
+    final ViewConfiguration viewConfigurationB = createViewConfiguration();
+
+    expect(viewConfigurationA == viewConfigurationB, true);
```
Comment (goderbauer) [inline packages/flutter/test/rendering/view_test.dart:54]: test `!=` cases ?

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/rendering/view.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter/lib/src/rendering/view.dart:42]: Done

File: packages/flutter/test/rendering/view_test.dart (hunk 2 above)
Comment (dnfield) [reply packages/flutter/test/rendering/view_test.dart:13]: This is also indirectly covered in the replace root layer test, but added a more explicit one to the == and hashcode test.

File: packages/flutter/lib/src/rendering/view.dart [hunk 4]
```diff
@@ -38,6 +38,18 @@ class ViewConfiguration {
...
+  bool operator ==(Object other) {
+    if (other.runtimeType != runtimeType)
+      return false;
+    return other is ViewConfiguration
+        && other.size == size
+        && other.devicePixelRatio == devicePixelRatio;
+  }
+
+  @override
+  int get hashCode => hashValues(size.hashCode, devicePixelRatio);
```
Comment (yjbanov) [inline packages/flutter/lib/src/rendering/view.dart:51]: IIRC `hashValues` already calls `.hashCode` on its arguments, so passing `size.hashCode` becomes `size.hashCode.hashCode`, which, while not wrong, is unnecessary.

**Review by yjbanov — APPROVED** [review]
LGTM if LGT @goderbauer 

lgtm

File: packages/flutter/lib/src/rendering/view.dart (hunk 4 above)
Comment (dnfield) [reply packages/flutter/lib/src/rendering/view.dart:51]: You are correct

**Pull request merged**